### PR TITLE
Allow parsing timezone without fully provided time

### LINKED ISSUE #49351: Regression in date parsing ("2016-11-30T12+01" accepted before ES 7
**Elasticsearch version** (`bin/elasticsearch --version`):

5.6.12 vs 74.2

**Plugins installed**: []
unrelated

**JVM version** (`java -version`):
openjdk version "1.8.0_232"

**OS version** (`uname -a` if on a Unix-like system):
centos8 container
Linux 044afd19e126 4.9.184-linuxkit #1 SMP Tue Jul 2 22:58:16 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:
This admittedly weird, incomplete date time string with a timezone offset:
`2016-11-30T12+01`
is accepted as a `date` by ES 5.6.12 but **fails** on ES 7.4.2 (probably any 7.x?) with:
```
HTTP/1.1 400 Bad Request

      "type" : "illegal_argument_exception",
      "reason" : "failed to parse date field [2016-11-30T12+01] with format [strict_date_optional_time||epoch_millis]",
``` 
```
curl -X PUT "elasticsearch:9200/my_index?pretty" -H 'Content-Type: application/json' --data-binary '{
  "mappings": {
    "doc": {
      "properties": {
        "date": {
          "type":   "date"
        }
      }
    }
  }
}'

curl -i -XPUT -H 'Content-type: application/json' http://elasticsearch:9200/my_index/doc/1 --data-binary  '{"date": "2016-11-30T12+01"}'
HTTP/1.1 201 Created
```

```
curl -X PUT "elasticsearch:9200/my_index?include_type_name=true&pretty" -H 'Content-Type: application/json' --data-binary '{            
"mappings": {
   "doc" : {
      "properties": {
        "date": {
          "type":   "date"
        }
      }
    }
  }
}'

curl -i -XPUT -H 'Content-type: application/json' http://elasticsearch:9200/my_index/doc/1?pretty --data-binary  '{"date": "2016-11-30T12+01"}'

HTTP/1.1 400 Bad Request
Warning: 299 Elasticsearch-7.4.2-2f90bbf7b93631e52bafb59b3b049cb44ec25e96 "[types removal] Specifying types in document index requests is deprecated, use the typeless endpoints instead (/{index}/_doc/{id}, /{index}/_doc, or /{index}/_create/{id})."
content-type: application/json; charset=UTF-8
content-length: 767

{
  "error" : {
    "root_cause" : [
      {
        "type" : "mapper_parsing_exception",
        "reason" : "failed to parse field [date] of type [date] in document with id '1'. Preview of field's value: '2016-11-30T12+01'"
      }
    ],
    "type" : "mapper_parsing_exception",
    "reason" : "failed to parse field [date] of type [date] in document with id '1'. Preview of field's value: '2016-11-30T12+01'",
    "caused_by" : {
      "type" : "illegal_argument_exception",
      "reason" : "failed to parse date field [2016-11-30T12+01] with format [strict_date_optional_time||epoch_millis]",
      "caused_by" : {
        "type" : "date_time_parse_exception",
        "reason" : "Failed to parse with all enclosed parsers"
      }
    }
  },
  "status" : 400
}
```
Btw same behavior if the `strict_date_optional_time` is used explicitly in the mapping, it's the default in both elasticsearch versions.
```
          "format": "strict_date_optional_time||epoch_millis"
```
**Provide logs (if relevant)**:

See output above, it just seems like a change in what the date parser accepts. Can that be fixed or is this by design?

Besides the regression and incompatibility this introduces for users upgrading, can this cause any in-place **upgrade** problems with ES 5 => 7?

### PR DESCRIPTION
**Pull request by pgomulka**
strict_date_optional_time in Joda allowed to provide just a timezone without fully provided time. So dates like `2018-01-01T00+01 , 2018-01-01T00:00+01 , 2018-01-01T00:00:00+01 , 2018-01-01T00:00:00.000+01` were allowed.
This should be fixed in java.time implementation too.
However, Joda was incorrectly parsing `2018-01-01T+01 ` when no time part was provided. It is because +01 which should be considered an offset, is parsed as a signed hour part.
java.time implementation does not suffer from this and correctly parses this date. We are not allowing this though, as this is not allowed be iso8601 spec.
In essence this PR is only fixing parsing for `2018-01-01T00+01`. Other formats were allowed already.
closes https://github.com/elastic/elasticsearch/issues/49351

### CODE REVIEW

**Comment by rjernst** [comment]
I'm not sure this is something we want to do. I don't think ISO 8601 allows for timezone without a time. I thought we were moving towards deprecating joda specific functionality? This especially seems a weird feature to support since based on your description it appears joda doesn't actually support this as it intended, hence why the dueling tests are commented out.

**Comment by mark-vieira** [comment]
FYI, you'll need to merge in the latest changes from `master` to fix these CI failure due to the new Java 13 build requirement.

**Comment by pgomulka** [comment]
@elasticmachine update branch

File: server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java [hunk 1]
```diff
@@ -46,9 +49,46 @@
...
     }
+    @BeforeClass
+    public static void checkJvmProperties(){
+        assert ("SPI,COMPAT".equals(System.getProperty("java.locale.providers")))
+            : "`-Djava.locale.providers=SPI,COMPAT` needs to be set";
+    }
+
+    public void testTimezoneParsing() {
+        //with Timezone
+//      assertSameDateAs("2016-11-30T+01", "strict_date_optional_time", "strict_date_optional_time");
```
Comment (pugnascotia) [inline server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java]: Do we perhaps want a comment here about Joda as well?

**Review by pugnascotia — APPROVED** [review]

**Comment by jamshid** [comment]
Thank you for fixing this! 
Both `2016-11-30T12+01` and `2016-11-30T+01` were parsed as valid dates with 6.8.x.
But only the former will continue to parse in 7.6.0?

**Comment by pgomulka** [comment]
@jamshid as per my comment here https://github.com/elastic/elasticsearch/pull/50178/files#diff-216a9f905247015f48298c3b6b1bb3a0R68
**our** `strict_date_optional_time` in Joda incorrectly parses `2016-11-30T+01` as `2016-11-30T01:00:00Z` when it should be `2016-11-29T23:00:00Z`. So even though you were not getting any exceptions, it was parsing these dates incorrectly
`date_optional_time` does not suffer from this 

please bear in mind, I need to confirm with ISO8601 if this should be merged at all.

**Comment by pgomulka** [comment]
@elasticmachine update branch

**Comment by pgomulka** [comment]
@rjernst it looks that this should indeed be allowed as `ISO/WD 8601-1` allows time representation with reduced accuracy
draft iso spec:
http://www.loc.gov/standards/datetime/iso-tc154-wg5_n0038_iso_wd_8601-1_2016-02-16.pdf
points 4.2.2.3 - 4.2.5.2.

**Comment by pgomulka** [comment]
I actually re-read the spec and it looks that timezone can only by applied with time. 
```
Expressions	 of	 the	 difference	 between	 local	 time	 and	 UTC	 of	 day	 are	 a	 component	 in	 the	
representations	defined	in	4.2.5.2;	they	shall	not	be	used	as	self-standing	expressions.
```
where 
```
4.2.5.2 Local	time	and	the	difference	from	UTC
```
allows reduced accuracy of local time. This is defined in 
```
4.2.2.3 Representations	with	reduced	accuracy
```
and allows only `hh:mm` or `hh`

Therefore 2016-11-30T12+01 is correct 
but 2016-11-30T+01 is incorrect. 
You can also see a discussion here https://stackoverflow.com/questions/52215961/how-can-i-write-an-iso-8601-date-with-a-timezone-but-no-time-component
I understand this was parsed successfully in 6.8 but I would consider this to be a bug in joda.

**Comment by pgomulka** [comment]
@elasticmachine update branch

**Review by pugnascotia — APPROVED** [review]

**Pull request merged**